src/lib.rs

```diff
@@ -74,49 +90,242 @@
+//!
+//! This should provide you with a trace of all the concurrency events leading up to the failure,
+//! which should allow you to identify how the bug is triggered.
+//!
+//! # Limitations and Caveats
+//!
+//! ## Large Models
+//!
+//! By default, loom runs an **exhaustive** check of your program's possible concurrent executions
+//! where **all** possible interleavings are checked. Loom's state reduction algorithms (see
+//! "Implementation" below) significantly reduce the state space that must be explored, but complex
+//! models can still take **significant** time to complete.
+//!
+//! To handle such large models in a more reasonable amount of time, you may need to **not** run
+//! an exhaustive check, and instead tell loom to prune out interleavings that are unlikely to
+//! reveal additional bugs. You do this by providing loom with a _thread pre-emption bound_. If you
+//! set such a bound, loom will check all possible executions that include **at most** `n` thread
+//! pre-emptions (where one thread is forcibly stopped and another one runs in its place. **In
+//! practice, setting the thread pre-emption bound to 2 or 3 is enough to catch most bugs** while
+//! significantly reducing the number of possible executions.
+//!
+//! To set the thread pre-emption bound, set the `LOOM_MAX_PREEMPTIONS` environment
+//! variable when running tests (or set
+//! [`Builder::preemption_bound`](model::Builder::preemption_bound)). For example:
+//!
+//! ```console
+//! LOOM_MAX_PREEMPTIONS=3 RUSTFLAGS="--cfg loom" cargo test --test loom_my_struct --release
+//! ```
+//!
+//! ## Relaxed Memory Ordering
+//!
+//! The [`Relaxed` memory ordering](std::sync::atomic::Ordering::Relaxed) allows particularly
+//! strange executions. For example, in the following code snippet, it is [completely
+//! legal][spec-relaxed] for `r1 == r2 == 42`!
+//!
+//! ```rust,no_run
+//! # use std::sync::atomic::AtomicUsize;
+//! # use std::sync::Ordering;
+//! # let x: &'static _ = Box::leak(Box::new(AtomicUsize::new(0)));
+//! # let y: &'static _ = Box::leak(Box::new(AtomicUsize::new(0)));
+//! thread::spawn(move || {
+//!   let r1 = y.load(Ordering::Relaxed); // A
+//!   x.store(r1, Ordering::Relaxed);     // B
+//! });
+//! thread::spawn(move || {
+//!   let r2 = x.load(Ordering::Relaxed); // C
+//!   y.store(42, Ordering::Relaxed);     // D
+//! });
+//! ```
+//!
+//! Unfortunately, it is not possible for loom to completely model all the interleavings that
+//! relaxed memory ordering allows. This is because the relaxed memory ordering allows memory
+//! operations to be re-ordered within a single thread -- B can run *before* A -- which loom cannot
+//! emulate. The same restriction applies to certain reorderings that are possible across different
+//! atomic variables with other memory orderings.
+//!
+//! Ultimately, this restriction means that loom is not a _completely_ exhaustive checker, and may
+//! miss bugs that could occur under particularly weak memory models.
```

Comment:
> miss bugs that could occur under particularly weak memory models.

This is not really clear. Loom checks the C++ memory model, which holds across all CPUs (that implement it). I'm *guessing* you are trying to say that loom may not catch bugs that can happen on ARM (CPU memory model), but this specific bug is probably an artifact of a compiler optimization and not the CPU's memory model.

Also, if one does not use `Relaxed`, then loom should be complete (modulo bugs, of which we know of one w/ SeqCst).

Reply:
The memory order affects both what the CPU is allowed to do _and_ what the compiler is allowed to do, so in this case I'm trying to get at the fact that the memory model allows a relaxed operations to appear to happen in different orders on different threads. How exactly that happens isn't all that important.

I think the same fundamental limitation comes up with `Acquire`+`Release` too in certain cases. Specifically, if a thread does a `store(Release)` on one variable followed by a `load(Acquire)` on another variable, I _think_ it is legal for the compiler or CPU to expose them in the opposite order to other threads, which I don't think loom can model the effects of?

Reply:
I ended up removing that last short paragraph. I think the paragraph above is sufficient.